# LOAD DATA

In [4]:
import os
print(dir)
def list_files_in_directory(directory_path):
    """List all files in the given directory."""
    try:
        files = os.listdir(directory_path)
        return [f for f in files if os.path.isfile(os.path.join(directory_path, f))]
    except FileNotFoundError:
        return f"Error: The directory '{directory_path}' does not exist."
    except PermissionError:
        return f"Error: Permission denied to access '{directory_path}'."
    
data = list_files_in_directory("../data-csv")
print(data)

<built-in function dir>
['costumer-churn.csv']


In [ ]:
import pandas as pd 
import numpy as np  


pdData = pd.read_csv("../data-csv/costumer-churn.csv")
print("read complete")

read complete


In [9]:
pdData.head(10)

customerID  gender  SeniorCitizen Partner Dependents  tenure PhoneService  \
0  7590-VHVEG  Female              0     Yes         No       1           No   
1  5575-GNVDE    Male              0      No         No      34          Yes   
2  3668-QPYBK    Male              0      No         No       2          Yes   
3  7795-CFOCW    Male              0      No         No      45           No   
4  9237-HQITU  Female              0      No         No       2          Yes   
5  9305-CDSKC  Female              0      No         No       8          Yes   
6  1452-KIOVK    Male              0      No        Yes      22          Yes   
7  6713-OKOMC  Female              0      No         No      10           No   
8  7892-POOKP  Female              0     Yes         No      28          Yes   
9  6388-TABGU    Male              0      No        Yes      62          Yes   

      MultipleLines InternetService OnlineSecurity  ... DeviceProtection  \
0  No phone service             DSL             No  ...               No   
1                No             DSL            Yes  ...              Yes   
2                No             DSL            Yes  ...               No   
3  No phone service             DSL            Yes  ...              Yes   
4                No     Fiber optic             No  ...               No   
5               Yes     Fiber optic             No  ...              Yes   
6               Yes     Fiber optic             No  ...               No   
7  No phone service             DSL            Yes  ...               No   
8               Yes     Fiber optic             No  ...              Yes   
9                No             DSL            Yes  ...               No   

  TechSupport StreamingTV StreamingMovies        Contract PaperlessBilling  \
0          No          No              No  Month-to-month              Yes   
1          No          No              No        One year               No   
2          No          No              No  Month-to-month              Yes   
3         Yes          No              No        One year               No   
4          No          No              No  Month-to-month              Yes   
5          No         Yes             Yes  Month-to-month              Yes   
6          No         Yes              No  Month-to-month              Yes   
7          No          No              No  Month-to-month               No   
8         Yes         Yes             Yes  Month-to-month              Yes   
9          No          No              No        One year               No   

               PaymentMethod MonthlyCharges  TotalCharges Churn  
0           Electronic check          29.85         29.85    No  
1               Mailed check          56.95        1889.5    No  
2               Mailed check          53.85        108.15   Yes  
3  Bank transfer (automatic)          42.30       1840.75    No  
4           Electronic check          70.70        151.65   Yes  
5           Electronic check          99.65         820.5   Yes  
6    Credit card (automatic)          89.10        1949.4    No  
7               Mailed check          29.75         301.9    No  
8           Electronic check         104.80       3046.05   Yes  
9  Bank transfer (automatic)          56.15       3487.95    No  

[10 rows x 21 columns]

In [10]:
pdData.describe()

SeniorCitizen       tenure  MonthlyCharges
count    7043.000000  7043.000000     7043.000000
mean        0.162147    32.371149       64.761692
std         0.368612    24.559481       30.090047
min         0.000000     0.000000       18.250000
25%         0.000000     9.000000       35.500000
50%         0.000000    29.000000       70.350000
75%         0.000000    55.000000       89.850000
max         1.000000    72.000000      118.750000

In [ ]:
pdData.info()


pdData.shape

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7043 entries, 0 to 7042
Data columns (total 21 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   customerID        7043 non-null   object 
 1   gender            7043 non-null   object 
 2   SeniorCitizen     7043 non-null   int64  
 3   Partner           7043 non-null   object 
 4   Dependents        7043 non-null   object 
 5   tenure            7043 non-null   int64  
 6   PhoneService      7043 non-null   object 
 7   MultipleLines     7043 non-null   object 
 8   InternetService   7043 non-null   object 
 9   OnlineSecurity    7043 non-null   object 
 10  OnlineBackup      7043 non-null   object 
 11  DeviceProtection  7043 non-null   object 
 12  TechSupport       7043 non-null   object 
 13  StreamingTV       7043 non-null   object 
 14  StreamingMovies   7043 non-null   object 
 15  Contract          7043 non-null   object 
 16  PaperlessBilling  7043 non-null   object 


(7043, 21)

# DATA PRE-PROCESSING

HANDLE CATEGORICAL DATA
Remove duplicates
Fix obvious invalid rows
Standardize column names
Convert datatypes (string → numeric/date)

In [12]:
pdData["gender"].isnull().sum()

np.int64(0)

In [13]:
pdData["gender"] = pdData["gender"].map({"Male": 0, "Female":1})
pdData.head()

customerID  gender  SeniorCitizen Partner Dependents  tenure PhoneService  \
0  7590-VHVEG       1              0     Yes         No       1           No   
1  5575-GNVDE       0              0      No         No      34          Yes   
2  3668-QPYBK       0              0      No         No       2          Yes   
3  7795-CFOCW       0              0      No         No      45           No   
4  9237-HQITU       1              0      No         No       2          Yes   

      MultipleLines InternetService OnlineSecurity  ... DeviceProtection  \
0  No phone service             DSL             No  ...               No   
1                No             DSL            Yes  ...              Yes   
2                No             DSL            Yes  ...               No   
3  No phone service             DSL            Yes  ...              Yes   
4                No     Fiber optic             No  ...               No   

  TechSupport StreamingTV StreamingMovies        Contract PaperlessBilling  \
0          No          No              No  Month-to-month              Yes   
1          No          No              No        One year               No   
2          No          No              No  Month-to-month              Yes   
3         Yes          No              No        One year               No   
4          No          No              No  Month-to-month              Yes   

               PaymentMethod MonthlyCharges  TotalCharges Churn  
0           Electronic check          29.85         29.85    No  
1               Mailed check          56.95        1889.5    No  
2               Mailed check          53.85        108.15   Yes  
3  Bank transfer (automatic)          42.30       1840.75    No  
4           Electronic check          70.70        151.65   Yes  

[5 rows x 21 columns]

In [14]:
pdData["Partner"].unique()

array(['Yes', 'No'], dtype=object)

In [15]:
pdData["Partner"] = pdData["Partner"].map({"Yes": 0, "No":1})
pdData.head()

customerID  gender  SeniorCitizen  Partner Dependents  tenure PhoneService  \
0  7590-VHVEG       1              0        0         No       1           No   
1  5575-GNVDE       0              0        1         No      34          Yes   
2  3668-QPYBK       0              0        1         No       2          Yes   
3  7795-CFOCW       0              0        1         No      45           No   
4  9237-HQITU       1              0        1         No       2          Yes   

      MultipleLines InternetService OnlineSecurity  ... DeviceProtection  \
0  No phone service             DSL             No  ...               No   
1                No             DSL            Yes  ...              Yes   
2                No             DSL            Yes  ...               No   
3  No phone service             DSL            Yes  ...              Yes   
4                No     Fiber optic             No  ...               No   

  TechSupport StreamingTV StreamingMovies        Contract PaperlessBilling  \
0          No          No              No  Month-to-month              Yes   
1          No          No              No        One year               No   
2          No          No              No  Month-to-month              Yes   
3         Yes          No              No        One year               No   
4          No          No              No  Month-to-month              Yes   

               PaymentMethod MonthlyCharges  TotalCharges Churn  
0           Electronic check          29.85         29.85    No  
1               Mailed check          56.95        1889.5    No  
2               Mailed check          53.85        108.15   Yes  
3  Bank transfer (automatic)          42.30       1840.75    No  
4           Electronic check          70.70        151.65   Yes  

[5 rows x 21 columns]

In [16]:
pdData["Dependents"].unique()

array(['No', 'Yes'], dtype=object)

In [17]:
pdData["Dependents"].isnull().sum()

np.int64(0)

In [18]:
pdData["Dependents"] = pdData["Dependents"].map({"No": 0, "Yes":1})
pdData.head()

customerID  gender  SeniorCitizen  Partner  Dependents  tenure  \
0  7590-VHVEG       1              0        0           0       1   
1  5575-GNVDE       0              0        1           0      34   
2  3668-QPYBK       0              0        1           0       2   
3  7795-CFOCW       0              0        1           0      45   
4  9237-HQITU       1              0        1           0       2   

  PhoneService     MultipleLines InternetService OnlineSecurity  ...  \
0           No  No phone service             DSL             No  ...   
1          Yes                No             DSL            Yes  ...   
2          Yes                No             DSL            Yes  ...   
3           No  No phone service             DSL            Yes  ...   
4          Yes                No     Fiber optic             No  ...   

  DeviceProtection TechSupport StreamingTV StreamingMovies        Contract  \
0               No          No          No              No  Month-to-month   
1              Yes          No          No              No        One year   
2               No          No          No              No  Month-to-month   
3              Yes         Yes          No              No        One year   
4               No          No          No              No  Month-to-month   

  PaperlessBilling              PaymentMethod MonthlyCharges  TotalCharges  \
0              Yes           Electronic check          29.85         29.85   
1               No               Mailed check          56.95        1889.5   
2              Yes               Mailed check          53.85        108.15   
3               No  Bank transfer (automatic)          42.30       1840.75   
4              Yes           Electronic check          70.70        151.65   

  Churn  
0    No  
1    No  
2   Yes  
3    No  
4   Yes  

[5 rows x 21 columns]

In [19]:
pdData["tenure"].unique()

array([ 1, 34,  2, 45,  8, 22, 10, 28, 62, 13, 16, 58, 49, 25, 69, 52, 71,
       21, 12, 30, 47, 72, 17, 27,  5, 46, 11, 70, 63, 43, 15, 60, 18, 66,
        9,  3, 31, 50, 64, 56,  7, 42, 35, 48, 29, 65, 38, 68, 32, 55, 37,
       36, 41,  6,  4, 33, 67, 23, 57, 61, 14, 20, 53, 40, 59, 24, 44, 19,
       54, 51, 26,  0, 39])

In [20]:
pdData["PhoneService"].unique()

array(['No', 'Yes'], dtype=object)

In [21]:
pdData["PhoneService"].isnull().sum()

np.int64(0)

In [22]:
pdData["PhoneService"] = pdData["PhoneService"].map({"No": 0, "Yes":1})
pdData.head()

customerID  gender  SeniorCitizen  Partner  Dependents  tenure  \
0  7590-VHVEG       1              0        0           0       1   
1  5575-GNVDE       0              0        1           0      34   
2  3668-QPYBK       0              0        1           0       2   
3  7795-CFOCW       0              0        1           0      45   
4  9237-HQITU       1              0        1           0       2   

   PhoneService     MultipleLines InternetService OnlineSecurity  ...  \
0             0  No phone service             DSL             No  ...   
1             1                No             DSL            Yes  ...   
2             1                No             DSL            Yes  ...   
3             0  No phone service             DSL            Yes  ...   
4             1                No     Fiber optic             No  ...   

  DeviceProtection TechSupport StreamingTV StreamingMovies        Contract  \
0               No          No          No              No  Month-to-month   
1              Yes          No          No              No        One year   
2               No          No          No              No  Month-to-month   
3              Yes         Yes          No              No        One year   
4               No          No          No              No  Month-to-month   

  PaperlessBilling              PaymentMethod MonthlyCharges  TotalCharges  \
0              Yes           Electronic check          29.85         29.85   
1               No               Mailed check          56.95        1889.5   
2              Yes               Mailed check          53.85        108.15   
3               No  Bank transfer (automatic)          42.30       1840.75   
4              Yes           Electronic check          70.70        151.65   

  Churn  
0    No  
1    No  
2   Yes  
3    No  
4   Yes  

[5 rows x 21 columns]

In [23]:
pdData.isnull().sum()

customerID          0
gender              0
SeniorCitizen       0
Partner             0
Dependents          0
tenure              0
PhoneService        0
MultipleLines       0
InternetService     0
OnlineSecurity      0
OnlineBackup        0
DeviceProtection    0
TechSupport         0
StreamingTV         0
StreamingMovies     0
Contract            0
PaperlessBilling    0
PaymentMethod       0
MonthlyCharges      0
TotalCharges        0
Churn               0
dtype: int64

In [25]:
pdData["MultipleLines"].unique()

array(['No phone service', 'No', 'Yes'], dtype=object)

In [26]:
pdData["MultipleLines"] = pdData["MultipleLines"].map({ "No phone service":0 , "No":1, "Yes":2})
pdData.head()

customerID  gender  SeniorCitizen  Partner  Dependents  tenure  \
0  7590-VHVEG       1              0        0           0       1   
1  5575-GNVDE       0              0        1           0      34   
2  3668-QPYBK       0              0        1           0       2   
3  7795-CFOCW       0              0        1           0      45   
4  9237-HQITU       1              0        1           0       2   

   PhoneService  MultipleLines InternetService OnlineSecurity  ...  \
0             0              0             DSL             No  ...   
1             1              1             DSL            Yes  ...   
2             1              1             DSL            Yes  ...   
3             0              0             DSL            Yes  ...   
4             1              1     Fiber optic             No  ...   

  DeviceProtection TechSupport StreamingTV StreamingMovies        Contract  \
0               No          No          No              No  Month-to-month   
1              Yes          No          No              No        One year   
2               No          No          No              No  Month-to-month   
3              Yes         Yes          No              No        One year   
4               No          No          No              No  Month-to-month   

  PaperlessBilling              PaymentMethod MonthlyCharges  TotalCharges  \
0              Yes           Electronic check          29.85         29.85   
1               No               Mailed check          56.95        1889.5   
2              Yes               Mailed check          53.85        108.15   
3               No  Bank transfer (automatic)          42.30       1840.75   
4              Yes           Electronic check          70.70        151.65   

  Churn  
0    No  
1    No  
2   Yes  
3    No  
4   Yes  

[5 rows x 21 columns]

In [27]:
pdData["InternetService"].unique()

array(['DSL', 'Fiber optic', 'No'], dtype=object)

In [28]:
pdData["InternetService"] = pdData["InternetService"].map({ "DSL":0 , "Fibre optic":1, "No":2})
pdData.head()

customerID  gender  SeniorCitizen  Partner  Dependents  tenure  \
0  7590-VHVEG       1              0        0           0       1   
1  5575-GNVDE       0              0        1           0      34   
2  3668-QPYBK       0              0        1           0       2   
3  7795-CFOCW       0              0        1           0      45   
4  9237-HQITU       1              0        1           0       2   

   PhoneService  MultipleLines  InternetService OnlineSecurity  ...  \
0             0              0              0.0             No  ...   
1             1              1              0.0            Yes  ...   
2             1              1              0.0            Yes  ...   
3             0              0              0.0            Yes  ...   
4             1              1              NaN             No  ...   

  DeviceProtection TechSupport StreamingTV StreamingMovies        Contract  \
0               No          No          No              No  Month-to-month   
1              Yes          No          No              No        One year   
2               No          No          No              No  Month-to-month   
3              Yes         Yes          No              No        One year   
4               No          No          No              No  Month-to-month   

  PaperlessBilling              PaymentMethod MonthlyCharges  TotalCharges  \
0              Yes           Electronic check          29.85         29.85   
1               No               Mailed check          56.95        1889.5   
2              Yes               Mailed check          53.85        108.15   
3               No  Bank transfer (automatic)          42.30       1840.75   
4              Yes           Electronic check          70.70        151.65   

  Churn  
0    No  
1    No  
2   Yes  
3    No  
4   Yes  

[5 rows x 21 columns]

In [29]:
pdData.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7043 entries, 0 to 7042
Data columns (total 21 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   customerID        7043 non-null   object 
 1   gender            7043 non-null   int64  
 2   SeniorCitizen     7043 non-null   int64  
 3   Partner           7043 non-null   int64  
 4   Dependents        7043 non-null   int64  
 5   tenure            7043 non-null   int64  
 6   PhoneService      7043 non-null   int64  
 7   MultipleLines     7043 non-null   int64  
 8   InternetService   3947 non-null   float64
 9   OnlineSecurity    7043 non-null   object 
 10  OnlineBackup      7043 non-null   object 
 11  DeviceProtection  7043 non-null   object 
 12  TechSupport       7043 non-null   object 
 13  StreamingTV       7043 non-null   object 
 14  StreamingMovies   7043 non-null   object 
 15  Contract          7043 non-null   object 
 16  PaperlessBilling  7043 non-null   object 


In [33]:
pdData["OnlineSecurity"].unique()

array([0, 1, 2])

In [31]:
pdData["OnlineSecurity"] = pdData["OnlineSecurity"].map({ "No":0 , "Yes":1, "No internet service":2})
pdData.head()

customerID  gender  SeniorCitizen  Partner  Dependents  tenure  \
0  7590-VHVEG       1              0        0           0       1   
1  5575-GNVDE       0              0        1           0      34   
2  3668-QPYBK       0              0        1           0       2   
3  7795-CFOCW       0              0        1           0      45   
4  9237-HQITU       1              0        1           0       2   

   PhoneService  MultipleLines  InternetService  OnlineSecurity  ...  \
0             0              0              0.0               0  ...   
1             1              1              0.0               1  ...   
2             1              1              0.0               1  ...   
3             0              0              0.0               1  ...   
4             1              1              NaN               0  ...   

  DeviceProtection TechSupport StreamingTV StreamingMovies        Contract  \
0               No          No          No              No  Month-to-month   
1              Yes          No          No              No        One year   
2               No          No          No              No  Month-to-month   
3              Yes         Yes          No              No        One year   
4               No          No          No              No  Month-to-month   

  PaperlessBilling              PaymentMethod MonthlyCharges  TotalCharges  \
0              Yes           Electronic check          29.85         29.85   
1               No               Mailed check          56.95        1889.5   
2              Yes               Mailed check          53.85        108.15   
3               No  Bank transfer (automatic)          42.30       1840.75   
4              Yes           Electronic check          70.70        151.65   

  Churn  
0    No  
1    No  
2   Yes  
3    No  
4   Yes  

[5 rows x 21 columns]

In [34]:
pdData["OnlineBackup"].unique()

array(['Yes', 'No', 'No internet service'], dtype=object)

In [35]:
pdData["OnlineBackup"] = pdData["OnlineBackup"].map({ "Yes":0 , "No":1, "No internet service":2})
pdData.head()

customerID  gender  SeniorCitizen  Partner  Dependents  tenure  \
0  7590-VHVEG       1              0        0           0       1   
1  5575-GNVDE       0              0        1           0      34   
2  3668-QPYBK       0              0        1           0       2   
3  7795-CFOCW       0              0        1           0      45   
4  9237-HQITU       1              0        1           0       2   

   PhoneService  MultipleLines  InternetService  OnlineSecurity  ...  \
0             0              0              0.0               0  ...   
1             1              1              0.0               1  ...   
2             1              1              0.0               1  ...   
3             0              0              0.0               1  ...   
4             1              1              NaN               0  ...   

   DeviceProtection TechSupport StreamingTV StreamingMovies        Contract  \
0                No          No          No              No  Month-to-month   
1               Yes          No          No              No        One year   
2                No          No          No              No  Month-to-month   
3               Yes         Yes          No              No        One year   
4                No          No          No              No  Month-to-month   

  PaperlessBilling              PaymentMethod MonthlyCharges  TotalCharges  \
0              Yes           Electronic check          29.85         29.85   
1               No               Mailed check          56.95        1889.5   
2              Yes               Mailed check          53.85        108.15   
3               No  Bank transfer (automatic)          42.30       1840.75   
4              Yes           Electronic check          70.70        151.65   

  Churn  
0    No  
1    No  
2   Yes  
3    No  
4   Yes  

[5 rows x 21 columns]

 
 12  TechSupport       7043 non-null   object 
 13  StreamingTV       7043 non-null   object 
 14  StreamingMovies   7043 non-null   object 
 15  Contract          7043 non-null   object 
 16  PaperlessBilling  7043 non-null   object 
 17  PaymentMethod     7043 non-null   object 
 18  MonthlyCharges    7043 non-null   float64
 19  TotalCharges      7043 non-null   object 
 20  Churn             7043 non-null   object 

In [36]:
pdData["DeviceProtection"].unique()

array(['No', 'Yes', 'No internet service'], dtype=object)

In [37]:
pdData["DeviceProtection"] = pdData["DeviceProtection"].map({ "No":0 , "Yes":1, "No internet service":2})
pdData.head()

customerID  gender  SeniorCitizen  Partner  Dependents  tenure  \
0  7590-VHVEG       1              0        0           0       1   
1  5575-GNVDE       0              0        1           0      34   
2  3668-QPYBK       0              0        1           0       2   
3  7795-CFOCW       0              0        1           0      45   
4  9237-HQITU       1              0        1           0       2   

   PhoneService  MultipleLines  InternetService  OnlineSecurity  ...  \
0             0              0              0.0               0  ...   
1             1              1              0.0               1  ...   
2             1              1              0.0               1  ...   
3             0              0              0.0               1  ...   
4             1              1              NaN               0  ...   

   DeviceProtection  TechSupport StreamingTV StreamingMovies        Contract  \
0                 0           No          No              No  Month-to-month   
1                 1           No          No              No        One year   
2                 0           No          No              No  Month-to-month   
3                 1          Yes          No              No        One year   
4                 0           No          No              No  Month-to-month   

  PaperlessBilling              PaymentMethod MonthlyCharges  TotalCharges  \
0              Yes           Electronic check          29.85         29.85   
1               No               Mailed check          56.95        1889.5   
2              Yes               Mailed check          53.85        108.15   
3               No  Bank transfer (automatic)          42.30       1840.75   
4              Yes           Electronic check          70.70        151.65   

  Churn  
0    No  
1    No  
2   Yes  
3    No  
4   Yes  

[5 rows x 21 columns]

In [38]:
pdData["TechSupport"].unique()

array(['No', 'Yes', 'No internet service'], dtype=object)

In [39]:
pdData["TechSupport"] = pdData["TechSupport"].map({ "No":0 , "Yes":1, "No internet service":2})
pdData.head()

customerID  gender  SeniorCitizen  Partner  Dependents  tenure  \
0  7590-VHVEG       1              0        0           0       1   
1  5575-GNVDE       0              0        1           0      34   
2  3668-QPYBK       0              0        1           0       2   
3  7795-CFOCW       0              0        1           0      45   
4  9237-HQITU       1              0        1           0       2   

   PhoneService  MultipleLines  InternetService  OnlineSecurity  ...  \
0             0              0              0.0               0  ...   
1             1              1              0.0               1  ...   
2             1              1              0.0               1  ...   
3             0              0              0.0               1  ...   
4             1              1              NaN               0  ...   

   DeviceProtection  TechSupport  StreamingTV StreamingMovies        Contract  \
0                 0            0           No              No  Month-to-month   
1                 1            0           No              No        One year   
2                 0            0           No              No  Month-to-month   
3                 1            1           No              No        One year   
4                 0            0           No              No  Month-to-month   

  PaperlessBilling              PaymentMethod MonthlyCharges  TotalCharges  \
0              Yes           Electronic check          29.85         29.85   
1               No               Mailed check          56.95        1889.5   
2              Yes               Mailed check          53.85        108.15   
3               No  Bank transfer (automatic)          42.30       1840.75   
4              Yes           Electronic check          70.70        151.65   

  Churn  
0    No  
1    No  
2   Yes  
3    No  
4   Yes  

[5 rows x 21 columns]

In [40]:
pdData.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7043 entries, 0 to 7042
Data columns (total 21 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   customerID        7043 non-null   object 
 1   gender            7043 non-null   int64  
 2   SeniorCitizen     7043 non-null   int64  
 3   Partner           7043 non-null   int64  
 4   Dependents        7043 non-null   int64  
 5   tenure            7043 non-null   int64  
 6   PhoneService      7043 non-null   int64  
 7   MultipleLines     7043 non-null   int64  
 8   InternetService   3947 non-null   float64
 9   OnlineSecurity    7043 non-null   int64  
 10  OnlineBackup      7043 non-null   int64  
 11  DeviceProtection  7043 non-null   int64  
 12  TechSupport       7043 non-null   int64  
 13  StreamingTV       7043 non-null   object 
 14  StreamingMovies   7043 non-null   object 
 15  Contract          7043 non-null   object 
 16  PaperlessBilling  7043 non-null   object 


In [41]:
pdData["StreamingTV"].unique()

array(['No', 'Yes', 'No internet service'], dtype=object)

In [42]:
pdData["StreamingTV"] = pdData["StreamingTV"].map({ "No":0 , "Yes":1, "No internet service":2})
pdData.head()

customerID  gender  SeniorCitizen  Partner  Dependents  tenure  \
0  7590-VHVEG       1              0        0           0       1   
1  5575-GNVDE       0              0        1           0      34   
2  3668-QPYBK       0              0        1           0       2   
3  7795-CFOCW       0              0        1           0      45   
4  9237-HQITU       1              0        1           0       2   

   PhoneService  MultipleLines  InternetService  OnlineSecurity  ...  \
0             0              0              0.0               0  ...   
1             1              1              0.0               1  ...   
2             1              1              0.0               1  ...   
3             0              0              0.0               1  ...   
4             1              1              NaN               0  ...   

   DeviceProtection  TechSupport  StreamingTV  StreamingMovies  \
0                 0            0            0               No   
1                 1            0            0               No   
2                 0            0            0               No   
3                 1            1            0               No   
4                 0            0            0               No   

         Contract PaperlessBilling              PaymentMethod MonthlyCharges  \
0  Month-to-month              Yes           Electronic check          29.85   
1        One year               No               Mailed check          56.95   
2  Month-to-month              Yes               Mailed check          53.85   
3        One year               No  Bank transfer (automatic)          42.30   
4  Month-to-month              Yes           Electronic check          70.70   

   TotalCharges Churn  
0         29.85    No  
1        1889.5    No  
2        108.15   Yes  
3       1840.75    No  
4        151.65   Yes  

[5 rows x 21 columns]

In [43]:
pdData["StreamingMovies"].unique()

array(['No', 'Yes', 'No internet service'], dtype=object)

In [44]:
pdData["StreamingMovies"] = pdData["StreamingMovies"].map({ "No":0 , "Yes":1, "No internet service":2})
pdData.head()

customerID  gender  SeniorCitizen  Partner  Dependents  tenure  \
0  7590-VHVEG       1              0        0           0       1   
1  5575-GNVDE       0              0        1           0      34   
2  3668-QPYBK       0              0        1           0       2   
3  7795-CFOCW       0              0        1           0      45   
4  9237-HQITU       1              0        1           0       2   

   PhoneService  MultipleLines  InternetService  OnlineSecurity  ...  \
0             0              0              0.0               0  ...   
1             1              1              0.0               1  ...   
2             1              1              0.0               1  ...   
3             0              0              0.0               1  ...   
4             1              1              NaN               0  ...   

   DeviceProtection  TechSupport  StreamingTV  StreamingMovies  \
0                 0            0            0                0   
1                 1            0            0                0   
2                 0            0            0                0   
3                 1            1            0                0   
4                 0            0            0                0   

         Contract PaperlessBilling              PaymentMethod MonthlyCharges  \
0  Month-to-month              Yes           Electronic check          29.85   
1        One year               No               Mailed check          56.95   
2  Month-to-month              Yes               Mailed check          53.85   
3        One year               No  Bank transfer (automatic)          42.30   
4  Month-to-month              Yes           Electronic check          70.70   

   TotalCharges Churn  
0         29.85    No  
1        1889.5    No  
2        108.15   Yes  
3       1840.75    No  
4        151.65   Yes  

[5 rows x 21 columns]

In [45]:
pdData["Contract"].unique()

array(['Month-to-month', 'One year', 'Two year'], dtype=object)

In [46]:
pdData["Contract"] = pdData["Contract"].map({ "Month-to-month":0 , "One year":1, "Two year":2})
pdData.head()

customerID  gender  SeniorCitizen  Partner  Dependents  tenure  \
0  7590-VHVEG       1              0        0           0       1   
1  5575-GNVDE       0              0        1           0      34   
2  3668-QPYBK       0              0        1           0       2   
3  7795-CFOCW       0              0        1           0      45   
4  9237-HQITU       1              0        1           0       2   

   PhoneService  MultipleLines  InternetService  OnlineSecurity  ...  \
0             0              0              0.0               0  ...   
1             1              1              0.0               1  ...   
2             1              1              0.0               1  ...   
3             0              0              0.0               1  ...   
4             1              1              NaN               0  ...   

   DeviceProtection  TechSupport  StreamingTV  StreamingMovies  Contract  \
0                 0            0            0                0         0   
1                 1            0            0                0         1   
2                 0            0            0                0         0   
3                 1            1            0                0         1   
4                 0            0            0                0         0   

   PaperlessBilling              PaymentMethod MonthlyCharges  TotalCharges  \
0               Yes           Electronic check          29.85         29.85   
1                No               Mailed check          56.95        1889.5   
2               Yes               Mailed check          53.85        108.15   
3                No  Bank transfer (automatic)          42.30       1840.75   
4               Yes           Electronic check          70.70        151.65   

  Churn  
0    No  
1    No  
2   Yes  
3    No  
4   Yes  

[5 rows x 21 columns]

In [47]:
pdData["PaperlessBilling"].head()

0    Yes
1     No
2    Yes
3     No
4    Yes
Name: PaperlessBilling, dtype: object

In [48]:
pdData["PaperlessBilling"].unique()

array(['Yes', 'No'], dtype=object)

In [49]:
pdData["PaperlessBilling"] = pdData["PaperlessBilling"].map({ "Yes":0 , "No":1})
pdData.head()

customerID  gender  SeniorCitizen  Partner  Dependents  tenure  \
0  7590-VHVEG       1              0        0           0       1   
1  5575-GNVDE       0              0        1           0      34   
2  3668-QPYBK       0              0        1           0       2   
3  7795-CFOCW       0              0        1           0      45   
4  9237-HQITU       1              0        1           0       2   

   PhoneService  MultipleLines  InternetService  OnlineSecurity  ...  \
0             0              0              0.0               0  ...   
1             1              1              0.0               1  ...   
2             1              1              0.0               1  ...   
3             0              0              0.0               1  ...   
4             1              1              NaN               0  ...   

   DeviceProtection  TechSupport  StreamingTV  StreamingMovies  Contract  \
0                 0            0            0                0         0   
1                 1            0            0                0         1   
2                 0            0            0                0         0   
3                 1            1            0                0         1   
4                 0            0            0                0         0   

   PaperlessBilling              PaymentMethod MonthlyCharges  TotalCharges  \
0                 0           Electronic check          29.85         29.85   
1                 1               Mailed check          56.95        1889.5   
2                 0               Mailed check          53.85        108.15   
3                 1  Bank transfer (automatic)          42.30       1840.75   
4                 0           Electronic check          70.70        151.65   

  Churn  
0    No  
1    No  
2   Yes  
3    No  
4   Yes  

[5 rows x 21 columns]

In [50]:
pdData["PaymentMethod"].unique()

array(['Electronic check', 'Mailed check', 'Bank transfer (automatic)',
       'Credit card (automatic)'], dtype=object)

In [ ]:
pdData["PaymentMethod"] = pdData["PaymentMethod"].map({ "Electronic check":0 , "Mailed check":1, "Bank transfer (automatic)":2, "Credit card (automatic)":3 })
pdData.head()



customerID  gender  SeniorCitizen  Partner  Dependents  tenure  \
0  7590-VHVEG       1              0        0           0       1   
1  5575-GNVDE       0              0        1           0      34   
2  3668-QPYBK       0              0        1           0       2   
3  7795-CFOCW       0              0        1           0      45   
4  9237-HQITU       1              0        1           0       2   

   PhoneService  MultipleLines  InternetService  OnlineSecurity  ...  \
0             0              0              0.0               0  ...   
1             1              1              0.0               1  ...   
2             1              1              0.0               1  ...   
3             0              0              0.0               1  ...   
4             1              1              NaN               0  ...   

   DeviceProtection  TechSupport  StreamingTV  StreamingMovies  Contract  \
0                 0            0            0                0         0   
1                 1            0            0                0         1   
2                 0            0            0                0         0   
3                 1            1            0                0         1   
4                 0            0            0                0         0   

   PaperlessBilling  PaymentMethod  MonthlyCharges  TotalCharges Churn  
0                 0            NaN           29.85         29.85    No  
1                 1            NaN           56.95        1889.5    No  
2                 0            NaN           53.85        108.15   Yes  
3                 1            NaN           42.30       1840.75    No  
4                 0            NaN           70.70        151.65   Yes  

[5 rows x 21 columns]

In [57]:
pdData.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7043 entries, 0 to 7042
Data columns (total 21 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   customerID        7043 non-null   object 
 1   gender            7043 non-null   int64  
 2   SeniorCitizen     7043 non-null   int64  
 3   Partner           7043 non-null   int64  
 4   Dependents        7043 non-null   int64  
 5   tenure            7043 non-null   int64  
 6   PhoneService      7043 non-null   int64  
 7   MultipleLines     7043 non-null   int64  
 8   InternetService   3947 non-null   float64
 9   OnlineSecurity    7043 non-null   int64  
 10  OnlineBackup      7043 non-null   int64  
 11  DeviceProtection  7043 non-null   int64  
 12  TechSupport       7043 non-null   int64  
 13  StreamingTV       7043 non-null   int64  
 14  StreamingMovies   7043 non-null   int64  
 15  Contract          7043 non-null   int64  
 16  PaperlessBilling  7043 non-null   int64  


In [ ]:
pdData["TotalCharges"].h

0      29.85
1     1889.5
2     108.15
3    1840.75
4     151.65
Name: TotalCharges, dtype: object

In [60]:
pdData["Churn"].unique()


array(['No', 'Yes'], dtype=object)

In [61]:
pdData["Churn"] = pdData["Churn"].map({ "No":0 , "Yes":1 }) 
pdData.head()



customerID  gender  SeniorCitizen  Partner  Dependents  tenure  \
0  7590-VHVEG       1              0        0           0       1   
1  5575-GNVDE       0              0        1           0      34   
2  3668-QPYBK       0              0        1           0       2   
3  7795-CFOCW       0              0        1           0      45   
4  9237-HQITU       1              0        1           0       2   

   PhoneService  MultipleLines  InternetService  OnlineSecurity  ...  \
0             0              0              0.0               0  ...   
1             1              1              0.0               1  ...   
2             1              1              0.0               1  ...   
3             0              0              0.0               1  ...   
4             1              1              NaN               0  ...   

   DeviceProtection  TechSupport  StreamingTV  StreamingMovies  Contract  \
0                 0            0            0                0         0   
1                 1            0            0                0         1   
2                 0            0            0                0         0   
3                 1            1            0                0         1   
4                 0            0            0                0         0   

   PaperlessBilling  PaymentMethod  MonthlyCharges  TotalCharges Churn  
0                 0            NaN           29.85         29.85     0  
1                 1            NaN           56.95        1889.5     0  
2                 0            NaN           53.85        108.15     1  
3                 1            NaN           42.30       1840.75     0  
4                 0            NaN           70.70        151.65     1  

[5 rows x 21 columns]

In [62]:
pdData[pdData["InternetService"].isna()]


customerID  gender  SeniorCitizen  Partner  Dependents  tenure  \
4     9237-HQITU       1              0        1           0       2   
5     9305-CDSKC       1              0        1           0       8   
6     1452-KIOVK       0              0        1           1      22   
8     7892-POOKP       1              0        0           0      28   
12    8091-TTVAX       0              0        0           0      58   
...          ...     ...            ...      ...         ...     ...   
7034  0639-TSIQW       1              0        1           0      67   
7035  8456-QDAVC       0              0        1           0      19   
7039  2234-XADUH       1              0        0           1      72   
7041  8361-LTMKD       0              1        0           0       4   
7042  3186-AJIEK       0              0        1           0      66   

      PhoneService  MultipleLines  InternetService  OnlineSecurity  ...  \
4                1              1              NaN               0  ...   
5                1              2              NaN               0  ...   
6                1              2              NaN               0  ...   
8                1              2              NaN               0  ...   
12               1              2              NaN               0  ...   
...            ...            ...              ...             ...  ...   
7034             1              2              NaN               1  ...   
7035             1              1              NaN               0  ...   
7039             1              2              NaN               0  ...   
7041             1              2              NaN               0  ...   
7042             1              1              NaN               1  ...   

      DeviceProtection  TechSupport  StreamingTV  StreamingMovies  Contract  \
4                    0            0            0                0         0   
5                    1            0            1                1         0   
6                    0            0            1                0         0   
8                    1            1            1                1         0   
12                   1            0            1                1         1   
...                ...          ...          ...              ...       ...   
7034                 1            0            1                0         0   
7035                 0            0            1                0         0   
7039                 1            0            1                1         1   
7041                 0            0            0                0         0   
7042                 1            1            1                1         2   

      PaperlessBilling  PaymentMethod  MonthlyCharges  TotalCharges Churn  
4                    0            NaN           70.70        151.65     1  
5                    0            NaN           99.65         820.5     1  
6                    0            NaN           89.10        1949.4     0  
8                    0            NaN          104.80       3046.05     1  
12                   1            NaN          100.35        5681.1     0  
...                ...            ...             ...           ...   ...  
7034                 0            NaN          102.95       6886.25     1  
7035                 0            NaN           78.70        1495.1     0  
7039                 0            NaN          103.20        7362.9     0  
7041                 0            NaN           74.40         306.6     1  
7042                 0            NaN          105.65        6844.5     0  

[3096 rows x 21 columns]

In [64]:
pdData.to_csv("Telco_Cleaned.csv", index=False)


# RESTORE THE ORIGINAL DATA IN THE CLEANED SET

In [65]:
backUpData = pdData
print("Backup complete")

Backup complete


In [67]:
originalcsv = pd.read_csv("../data-csv/costumer-churn.csv")

In [68]:
pdData["PaymentMethod"] = originalcsv["PaymentMethod"]

In [69]:
pdData["PaymentMethod"].head(15)

0              Electronic check
1                  Mailed check
2                  Mailed check
3     Bank transfer (automatic)
4              Electronic check
5              Electronic check
6       Credit card (automatic)
7                  Mailed check
8              Electronic check
9     Bank transfer (automatic)
10                 Mailed check
11      Credit card (automatic)
12      Credit card (automatic)
13    Bank transfer (automatic)
14             Electronic check
Name: PaymentMethod, dtype: object

In [70]:
pdData["PaymentMethod"] = pdData["PaymentMethod"].map({ "Electronic check":0 , "Mailed check":1, "Bank transfer (automatic)":2, "Credit card (automatic)":3 })
pdData.head()

customerID  gender  SeniorCitizen  Partner  Dependents  tenure  \
0  7590-VHVEG       1              0        0           0       1   
1  5575-GNVDE       0              0        1           0      34   
2  3668-QPYBK       0              0        1           0       2   
3  7795-CFOCW       0              0        1           0      45   
4  9237-HQITU       1              0        1           0       2   

   PhoneService  MultipleLines  InternetService  OnlineSecurity  ...  \
0             0              0              0.0               0  ...   
1             1              1              0.0               1  ...   
2             1              1              0.0               1  ...   
3             0              0              0.0               1  ...   
4             1              1              NaN               0  ...   

   DeviceProtection  TechSupport  StreamingTV  StreamingMovies  Contract  \
0                 0            0            0                0         0   
1                 1            0            0                0         1   
2                 0            0            0                0         0   
3                 1            1            0                0         1   
4                 0            0            0                0         0   

   PaperlessBilling  PaymentMethod  MonthlyCharges  TotalCharges Churn  
0                 0              0           29.85         29.85     0  
1                 1              1           56.95        1889.5     0  
2                 0              1           53.85        108.15     1  
3                 1              2           42.30       1840.75     0  
4                 0              0           70.70        151.65     1  

[5 rows x 21 columns]

In [71]:
pdData["InternetService"] = originalcsv["InternetService"]

In [ ]:
pdData["InternetService"].head(20)

In [ ]:
pdData["InternetService"] = pdData["InternetService"].map({ "DSL":0 , "Fibre optic":1, "No":2})
pdData.head()